File: scalechain-net/src/main/scala/io/scalechain/blockchain/net/Node.scala
package io.scalechain.blockchain.net

import com.typesafe.scalalogging.Logger
import io.scalechain.blockchain.chain.Blockchain
import io.scalechain.blockchain.net.message.GetBlocksFactory
import io.scalechain.blockchain.proto.{Hash, ProtocolMessage}
import io.scalechain.util.Config
import org.slf4j.LoggerFactory

object Node {
  var theNode : Node = null
  def create(peerCommunicator: PeerCommunicator, chain : Blockchain) : Node = {
    if (theNode == null) {
      theNode = new Node(peerCommunicator, chain)
    }
    theNode
  }
  def get() : Node = {
    assert( theNode != null )
    theNode
  }
}

/**
  * Created by kangmo on 8/8/16.
  */
class Node(peerCommunicator: PeerCommunicator, chain : Blockchain) {
  private val logger = Logger( LoggerFactory.getLogger(classOf[Node]) )

  // BUGBUG : What if the best peer was a malicious node?
  // The best peer for initial block download
  protected[net] var bestPeerForIBD : Peer = null

  protected[net] var lastBlockHashForIBD : Hash = Hash.ALL_ZERO

  def setLastBlockHashForIBD(blockHash : Hash) : Unit = {
    lastBlockHashForIBD = blockHash
  }
  def getLastBlockHashForIBD() : Hash = {
    lastBlockHashForIBD
  }

  /**
    * Return the best height if we received version message more than two thirds of total nodes.
    *
    * @return Some(best Peer) that has the highest Version.startHeight if we received version message more than two thirds of total nodes. None otherwise.
    */
  def getBestPeer() : Option[Peer] = {
    val maxPeerCount = Config.peerAddresses.length
    val peerInfos = peerCommunicator.getPeerInfos()

    // We have two way communication channels for each peer.
    // PeerInfos list each communication channel.
    val connectedPeers = peerInfos.length / 2

    // Get the total active peers including me.
    val totalActivePeers = connectedPeers + 1

    // Do we have enough number of peers? (At least more than two thirds)
    if ( totalActivePeers > maxPeerCount * 2 / 3 ) {
      // Did we receive starting height for all peers?
      if ( peerInfos.filter( _.startingheight.isDefined ).length == peerInfos.length ) {
        val bestPeer = peerCommunicator.getBestPeer.get
        Some(bestPeer)
      } else {
        None
      }
    } else { // Not enough connected peers.
      None
    }
  }

  def updateStatus(): Unit = {
    synchronized {
      val bestPeerOption = getBestPeer()
      if (bestPeerOption.isDefined) {
        val bestPeer = bestPeerOption.get
        if (bestPeer.versionOption.map(_.startHeight).getOrElse(0) > chain.getBestBlockHeight() ) {
          bestPeerForIBD = bestPeer

          /* Start IBD(Initial block download). Request getblocks message to get inv messages for our missing blocks */
          val getBlocksMessage = GetBlocksFactory.create(getLastBlockHashForIBD)
          bestPeerForIBD.send(getBlocksMessage)

          logger.info("Initial block download started. Requested getblocks message.")
        }
      }
    }
  }

  def bestPeerStartHeight() : Int = {
    synchronized {
      assert(bestPeerForIBD != null)
      bestPeerForIBD.versionOption.map(_.startHeight).getOrElse(0)
    }
  }

  def sendToBestPeer(message : ProtocolMessage) : Unit = {
    synchronized {
      assert(bestPeerForIBD != null)
      bestPeerForIBD.send(message)
    }
  }

  def isInitialBlockDownload() : Boolean = {
    synchronized {
      bestPeerForIBD != null
    }
  }

  def stopInitialBlockDownload(): Unit = {
    synchronized {
      bestPeerForIBD = null
    }
  }
}

File: scalechain-net/src/main/kotlin/io/scalechain/blockchain/net/Node.kt
package io.scalechain.blockchain.net

import io.scalechain.blockchain.chain.Blockchain
import io.scalechain.blockchain.net.message.GetBlocksFactory
import io.scalechain.blockchain.proto.Hash
import io.scalechain.blockchain.proto.ProtocolMessage
import io.scalechain.util.Config
import org.slf4j.LoggerFactory


/**
 * Created by kangmo on 8/8/16.
 */
class Node(private val peerCommunicator: PeerCommunicator, private val chain : Blockchain) {
    private val logger = LoggerFactory.getLogger(Node::class.java)

    // BUGBUG : What if the best peer was a malicious node?
    // The best peer for initial block download
    var bestPeerForIBD : Peer? = null

    var lastBlockHashForIBD : Hash = Hash.ALL_ZERO

    /**
     * Return the best height if we received version message more than two thirds of total nodes.
     *
     * @return Some(best Peer) that has the highest Version.startHeight if we received version message more than two thirds of total nodes. None otherwise.
     */
    fun getBestPeer() : Peer? {
        val maxPeerCount = Config.peerAddresses().size
        val peerInfos = peerCommunicator.getPeerInfos()

        // We have two way communication channels for each peer.
        // PeerInfos list each communication channel.
        val connectedPeers = peerInfos.size / 2

        // Get the total active peers including me.
        val totalActivePeers = connectedPeers + 1

        // Do we have enough number of peers? (At least more than two thirds)
        if ( totalActivePeers > maxPeerCount * 2 / 3 ) {
            // Did we receive starting height for all peers?
            if ( peerInfos.filter{ it.startingheight != null }.size == peerInfos.size ) {
                val bestPeer = peerCommunicator.getBestPeer()
                return bestPeer
            } else {
                return null
            }
        } else { // Not enough connected peers.
            return null
        }
    }

    fun updateStatus(): Unit {
        synchronized(this) {
            val bestPeerOption = getBestPeer()
            if (bestPeerOption != null) {
                val bestPeer = bestPeerOption
                if ( (bestPeer.versionOption?.startHeight ?: 0) > chain.getBestBlockHeight() ) {
                    bestPeerForIBD = bestPeer

                    /* Start IBD(Initial block download). Request getblocks message to get inv messages for our missing blocks */
                    val getBlocksMessage = GetBlocksFactory.create(lastBlockHashForIBD)
                    bestPeerForIBD!!.send(getBlocksMessage)

                    logger.info("Initial block download started. Requested getblocks message.")
                }
            }
        }
    }

    fun bestPeerStartHeight() : Int {
        synchronized(this) {
            assert(bestPeerForIBD != null)
            return bestPeerForIBD?.versionOption?.startHeight ?: 0
        }
    }

    fun sendToBestPeer(message : ProtocolMessage) : Unit {
        synchronized(this) {
            assert(bestPeerForIBD != null)
            bestPeerForIBD!!.send(message)
        }
    }

    fun isInitialBlockDownload() : Boolean {
        synchronized(this) {
            return bestPeerForIBD != null
        }
    }

    fun stopInitialBlockDownload(): Unit {
        synchronized(this) {
            bestPeerForIBD = null
        }
    }

    companion object {
        var theNode : Node? = null
        fun create(peerCommunicator: PeerCommunicator, chain : Blockchain) : Node {
            if (theNode == null) {
                theNode = Node(peerCommunicator, chain)
            }
            return theNode!!
        }
        fun get() : Node {
            assert( theNode != null )
            return theNode!!
        }
    }
}
